src/work/BasicWork.h

```diff
@@ -0,0 +1,114 @@
+ *  - May need to be broken into steps, so as not to block the main thread
+ *  - May fail and need to retry, after some delay
+ *
+ *  BasicWork manages all state transitions via `crankWork` as well as retries.
+ *  While customers can trigger cranking and check on BasicWork's status,
+ *  it is implementers' responsibility to implement `onRun`, that
+ *  hints the next desired state.
+ */
+class BasicWork : public std::enable_shared_from_this<BasicWork>,
+                  private NonMovableOrCopyable
+{
+  public:
+    static size_t const RETRY_NEVER;
+    static size_t const RETRY_ONCE;
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    enum State
+    {
+        WORK_RUNNING,
+        WORK_SUCCESS,
+        WORK_WAITING,
+        WORK_FAILURE_RETRY,
+        WORK_FAILURE_RAISE,
+        WORK_DESTRUCTING
+    };
+
+    BasicWork(Application& app, std::function<void()> callback,
+              std::string name, size_t maxRetries);
+    virtual ~BasicWork();
+
+    std::string const& getName() const;
+    virtual std::string getStatus() const;
+    State getState() const;
+    bool isDone() const;
+
+    size_t getMaxRetries() const;
+    uint64_t getRetryETA() const;
+
+    // Main method for state transition, mostly dictated by `onRun`
+    void crankWork();
+
+    // TODO (mlo) look into uses of reset in current interface
+    // Proper cleanup needs to happen at some point, and `reset`
+    // might not be the best place. For instance, `GetRemoteFileWork`
+    // calls std::remove is `onReset`, so an attempt to remove a non-existent
+    // file happens every time a child is added, which is inefficient.
+    void reset();
+
+  protected:
+    // Implementers can override these callbacks to customize functionality
+    // `onReset` is called on any important state transition, e.g. when new work
+    // is added, on a retry, success and failure. This method ensures that
+    // all the references (including children) are properly cleaned up for
+    // safe destruction. `onRun` performs necessary work logic, and hints the
+    // next state transition. `onSuccess` is called when work transitions into
+    // WORK_SUCCESS state. Similarly, `onFailure*` methods are called upon
+    // transitioning into appropriate failure state. An example usage would be
+    // cleaning up side effects, like downloaded files in these methods. If you
+    // want to force failure (and reset / retry) you can return
+    // WORK_FAILURE_RETRY or WORK_FAILURE_RAISE. After a retry count is
+    // passed, WORK_FAILURE_RETRY means WORK_FAILURE_RAISE anyways.
+    // WORK_FAILURE_FATAL is equivalent to WORK_FAILURE_RAISE passed up in
+    // the work chain - when WORK_FAILURE_FATAL is raised in one work item,
+    // all work items that leaded to this one will also fail without retrying.
+
+    virtual void onReset();
+    virtual State onRun() = 0;
+    virtual void onFailureRetry();
```

Comment:
do we really need the `onFailureXXX` methods? it seems that this creates an arbitrary split of the `onRun` logic.

Reply:
do we also need `onSuccess` I am not sure what problem this is solving?

Reply:
It's convenient for things like cleaning up side effects, marking metrics, logging, reporting progress (these are some of the use cases today). I don't see a problem having these -- in practice, `onRun` already seems to be doing a lot (keeping track of its work and in some cases of its children).